### DISCUSSION TITLE
Feature Request: add metadata in displaying files & directories (plugin system)

### DISCUSSION THREAD

**Discussion by hacker-DOM** [sxyazi/yazi#254] [4. Ideas]
Hi @community

Thaanks for the cool tool (I saw it a few days ago on Reddit and downloaded it now). I have been using ranger until now with two columns, but I would switch on a heartbeat (ranger freezes 20% of the time I use it). I have not been using Nnnn because it only supports one column, so I'm glad Yazi supports multiple out of the box

I am under the impression that there isn't a plugin system yet. I just wanted to voice a feature request that would be necessary and sufficient for me to move Ranger → Yazi.

I am a security auditor and I read a lot of code that I'm unfamiliar with. For that purpose, I created https://github.com/hacker-DOM/zsh-sccd, which counts lines of code in every directory and creates a tiny `.metadata.json` in every folder. I then display the line counts in my file managers, eg Neotree and Ranger.

I would switch to Yazi if I could create a plugin that would display these line numbers

Ideas: the plugin system could be in Lua or be stdin/stoud based like [pandoc](https://github.com/jgm/pandoc).

Neotree:
<img width="788" alt="image" src="https://github.com/sxyazi/yazi/assets/18601956/b1391597-e941-4ae6-ac05-fc7b2d18adb4">

Ranger:
<img width="668" alt="image" src="https://github.com/sxyazi/yazi/assets/18601956/d2279d04-1c28-4f20-8d66-3d1a679a9f27">

**Comment by sxyazi** [comment]
Actually, I think the current [theme system](https://github.com/sxyazi/yazi/pull/161) already supports it. It allows for most of the UI to be rewritten using Lua.

You can read `.metadata.json` during the initial UI rendering and store it in a variable. Then, on subsequent renderings, if the variable is not empty, you can use it directly without any IO operations. Since `.metadata.json` is tiny, this should be fast.

The later plugin system will extract this IO operation into a separate asynchronous pre-caching runtime, ensuring there is no UI blocking.
